M2/Macaulay2/packages/Hadamard.m2

```diff
@@ -0,0 +1,441 @@
+    if #delete(null, P) =!= #P then return "error: nuul entries are not allowed"
+    else new Point from P)
+
+makeListOfPoints=method()
+makeListOfPoints(VisibleList):=(L)->(
+             apply(L,P-> point P)
+	     )
+
+Point * Point:=(p,q)->(
+    if #p =!= #q then error("the points should be in a same projective space") else
+    pp:=apply(p,q,times);
+    if all(pp, i->i==0) then error("this operation is not well-defined in projective space")
+    else return pp
+    )
+
+Point == Point := (p,q)->(
+    rank pointsToMatrix({p,q}) == 1
+    )
+
+---Hadamard product of two points---
+hadProdPoints = method()	 
+hadProdPoints(Point,Point):=(p,q)->(
+     p * q
+    )
+
+hadProdPoints(List,List):=(p,q)->(
+ point p * point q
+    )
+
+hadProdPoints(Array,Array):=(p,q)->(
+   point p * point q
+    )
+
+-- Hadamrd product of two varieties
+hadProdOfVariety = method()
+hadProdOfVariety (Ideal, Ideal):= (I,J) -> (
+    newy := symbol newy;
+    newz := symbol newz;
+    varI:= gens ring I;
+    varJ:= gens ring J;
+    CRI:=coefficientRing ring I;
+    CRJ:=coefficientRing ring J;
+    RYI:=CRI[newy_0.. newy_(# varI -1)];
+    RZJ:=CRI[newz_0..newz_(# varJ -1)];
+    IY:=sub(I,apply(varI,gens RYI,(a,b)->a=>b));
+    JZ:=sub(J,apply(varJ,gens RZJ,(a,b)->a=>b));
+    TensorRingIJ:=RYI/IY ** RZJ/JZ;
+    use TensorRingIJ;
+    Projvars:=apply(#(gens ring I),i->newy_i * newz_i);
+    hadMap:=map(TensorRingIJ,ring I, Projvars);
+    ker hadMap
+    )
+
+
+-------Hadmard product of two subsets of points on two varieties---------
+hadProdListsOfPoints = method()
+hadProdListsOfPoints(List,List) :=(X,Y)->(
+     convert:= obj -> if not instance(obj, Point) then point(obj) else obj;
+     newX:=apply(X,convert);
+     newY:=apply(Y,convert);
+     return delete(null, for I in toList(set(newX) ** set(newY)) list (
+	 try(I_0 * I_1) then (I_0 * I_1) else null
+	 ))
+     )
+
+
+pointsToMatrix=method()
+pointsToMatrix(List):= (PTM) ->( matrix apply(PTM, toList))
+
+---Hadamard powers of varieties------------
+
+hadamardPower = method()
+hadamardPower(Ideal,ZZ):=(I,r)->(
+    if r<1 then return error("the second argument should be positve integre >=1 ");
+   NewI := I;
+   for i from 1 to r-1 do NewI = hadProdOfVariety(NewI,I);
+   return NewI)
+
+---Hadamard powers of sets of points ------------
+
+hadamardPower(List,ZZ):=(L,r)->(
+    if r<1 then return error("the second argument should be a positve integre >=1");
+   NewL := L;
+   for i from 1 to r-1 do NewL = hadProdListsOfPoints(NewL,L);
+   return toList set NewL)
+
+
+-----------------%%%--------------------------------%%%---------------
+
+hadamardMult=method()
+hadamardMult(List):=(L)->(
+    if not uniform L then
+     (return "error: entries should be in the same class");
+    if instance(first L,Ideal)==true then fold(hadProdOfVariety,L)
+    else
+    if instance(first L,List)==true or instance(first L,Point)==true then fold(hadProdPoints,L)
+    else return ("error: check the inputs")
```

Comment:
The error message should tell the user what is expected.

Reply:
Fixed